cpp/src/arrow/json/parser.cc

```diff
@@ -0,0 +1,1210 @@
+class AdaptiveListBuilder;
+
+template <typename T, typename... A>
+std::unique_ptr<T> make_unique(A&&... args) {
+  return std::unique_ptr<T>(new T(std::forward<A>(args)...));
+}
+
+class AdaptiveNullBuilder : public AdaptiveArrayBuilder {
+ public:
+  struct value_type {};
+
+  AdaptiveNullBuilder(MemoryPool* pool)
+      : AdaptiveArrayBuilder(initial_type()), pool_(pool) {}
+
+  Status Append(value_type) { return AppendNull(); }
+
+  Status AppendNull() {
+    ++length_;
+    return Status::OK();
+  }
+
+  Status Finish(std::shared_ptr<Array>* out) override {
+    *out = std::make_shared<NullArray>(length_);
+    return Status::OK();
+  }
+
+  Status MaybePromoteTo(std::shared_ptr<DataType> type,
+                        std::unique_ptr<AdaptiveArrayBuilder>* out) override;
+
+  static std::shared_ptr<DataType> initial_type() { return null(); }
+
+  static Status Make(MemoryPool* pool, int64_t leading_nulls,
+                     std::unique_ptr<AdaptiveArrayBuilder>* out) {
+    *out = make_unique<AdaptiveNullBuilder>(pool);
+    auto builder = static_cast<AdaptiveNullBuilder*>(out->get());
+    builder->length_ = leading_nulls;
+    return Status::OK();
+  }
+
+ private:
+  MemoryPool* pool_;
+};
+
+class AdaptiveBooleanBuilder : public AdaptiveArrayBuilder {
+ public:
+  using value_type = bool;
+
+  AdaptiveBooleanBuilder(MemoryPool* pool)
+      : AdaptiveArrayBuilder(initial_type()), boolean_builder_(pool) {}
+
+  Status Append(bool value) {
+    ++length_;
+    return boolean_builder_.Append(value);
+  }
+
+  Status AppendNull() {
+    ++length_;
+    return boolean_builder_.AppendNull();
+  }
+
+  Status Finish(std::shared_ptr<Array>* out) override {
+    return boolean_builder_.Finish(out);
+  }
+
+  Status MaybePromoteTo(std::shared_ptr<DataType> type,
+                        std::unique_ptr<AdaptiveArrayBuilder>*) override {
+    if (type->Equals(type_)) return Status::OK();
+    return ConversionError();
+  }
+
+  static std::shared_ptr<DataType> initial_type() { return boolean(); }
+
+  static Status Make(MemoryPool* pool, int64_t leading_nulls,
+                     std::unique_ptr<AdaptiveArrayBuilder>* out) {
+    *out = make_unique<AdaptiveBooleanBuilder>(pool);
+    auto builder = static_cast<AdaptiveBooleanBuilder*>(out->get());
+    builder->length_ = leading_nulls;
+    return builder->boolean_builder_.SetLeadingNulls(leading_nulls);
+  }
+
+ private:
+  struct WrappedBuilder : BooleanBuilder {
+    using BooleanBuilder::BooleanBuilder;
+
+    Status SetLeadingNulls(int64_t leading_nulls) {
+      ARROW_CHECK(length_ == 0);
+      null_count_ = leading_nulls;
+      length_ = leading_nulls;
+      return Resize(leading_nulls);
+    }
+  } boolean_builder_;
+};
+
+class Int64OrDoubleBuilder : public AdaptiveArrayBuilder {
+ public:
+  using value_type = string_view;
+
+  Int64OrDoubleBuilder(MemoryPool* pool)
+      : AdaptiveArrayBuilder(initial_type()), bytes_builder_(pool) {}
+
+  Status Append(string_view repr) {
```

Comment:
Tricky case: is the string "-0" treated consistently?

Reply:
Hmmm. If that were encountered before a cast to double, then we'd get `double(int64_t(-0))` which would be `double(+0)`. if it's encountered afterwards then we'd get `double(-0)`. Guess not